WIP: MIT CCX student dashboard display

cewing opened the pull request:
This pull request addresses display issues identified after the merging of the MIT CCX extension work in https://github.com/edx/edx-platform/pull/6636 and reported in [this JIRA ticket](https://openedx.atlassian.net/browse/TNL-2007) (and reported also as jazkarta/edx-platform#74).  At the same time, it also addresses an issue identified by the MIT team and reported as jazkarta/edx-platform#75.  

The PR updates the display of CCX courses to which a student has been granted membership to match the display of "normal" EdX courses.  It ensures that the start and end dates displayed for the CCX are in fact those of the CCX itself, and not those of the course from which it was created.  It also ensures that if the start date of a CCX has not yet arrived the student will not be able to click on the CCX and view it.

A CCX which has not yet started:
![not_yet_started](https://cloud.githubusercontent.com/assets/417109/7264287/ff095d5a-e83e-11e4-812b-017c3df7e562.png)


A CCX which has started:
![already_started](https://cloud.githubusercontent.com/assets/417109/7264294/081cd6a6-e83f-11e4-8606-d91cb5ae35b4.png)


The code updated for this pull request resides in the `lms/djangoapps/ccx` app as well as in the template for the shared dashboard.

This pull request is a Work In Progress, we seek advice on the approach taken to solving the problem of display of CCX instances on the shared dashboard as well as pointers to any opportunities to better use internal code to accomplish our goals. In particular, we are curious as to the advisability of replicating API provided by a Course Descriptor as part of the CCX model.  Is there a better way of accomplishing the goal of treating a CCX more similarly to a standard course?

For context on ongoing work to make CCXs work more like regular courses it may be valuable to look over the discussion on jazkarta/edx-platform#43 and the related tickets in that tracker which have been opened.  It may be that work yet to be done for that issue may help to resolve some of the questions asked here.

Please note that as this is a WIP PR, tests have yet to be written to cover the work done herein. They will be added once the direction is clear.

edx-webhook commented:
Thanks for the pull request, @cewing! I've created [OSPR-542](https://openedx.atlassian.net/browse/OSPR-542) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams. 

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ('this must be merged by XX date', and why that is)
- partner information ('this is a course on edx.org')
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the Github pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

sarina commented:
add to whitelist

sarina commented:
@cewing this is a WIP - please ping me when it's ready for review.

cewing commented:
@sarina: could you let me know who to ping to review the PR regarding answers to my questions in the description above?  To clarify there are a couple of things I'm not totally comfy with here, and would like feedback if possible:

1.  The advisability of reproducing course descriptor APIs on the CCX object (or suggestions of ways to avoid this, if that would be a better idea)
2.  The approach of keeping the display template for a CCX on the dashboard as a separate file as opposed to trying to merge CCX display with course display through the `_dashboard_course_listing.html` template.

sarina commented:
@nedbat can you please help Cris answer the questions in this PR, ASAP?

sarina commented:
@cewing I will say it would make a ton of sense to try to keep the templates separate, but yes, syncing styles will be a pain I fear.

cewing commented:
@sarina, @nedbat yes, my sense was that trying to inline ccx display into the course display template just seemed insane, but it does mean more work down the line.  The closer we can make CCXs behave to courses, the better, I think.  Might eventually be able to close the gap

cewing commented:
@sarina, I think this is ready to go.  It passes quality checks on my local machine and is now pretty well covered with tests.

sarina commented on lms/djangoapps/ccx/models.py:
```diff
@@ -16,6 +26,71 @@ class CustomCourseForEdX(models.Model):
+    _due = _MARKER
+
+    @property
+    def course(self):
```
Why do none of these have docstrings?

cewing replied:
because I'm a terrible person :)
Fixing now.

sarina commented:
@cewing I see a lot of pylint errors relating to missing docstrings

nedbat commented on lms/djangoapps/ccx/tests/test_models.py:
```diff
@@ -125,3 +131,160 @@ def test_future_membership_disallows_auto_enroll(self):
+        self.set_ccx_override('start', expected)
+        actual = self.ccx.start
+        diff = expected - actual
+        self.assertEqual(diff.seconds, 0)
```
Not to be pedantic, but the .days field could be non-zero.  .total_second() might be better to use.

cewing replied:
thanks, will fix.

nedbat commented on lms/djangoapps/ccx/tests/test_models.py:
```diff
@@ -125,3 +131,160 @@ def test_future_membership_disallows_auto_enroll(self):
+        """verify that due returns None when the field has not been set"""
+        expected = None
+        actual = self.ccx.due
+        self.assertTrue(expected is actual)
```
`self.assertIsNone` would work well here.

nedbat commented on lms/djangoapps/ccx/tests/test_models.py:
```diff
@@ -125,3 +131,160 @@ def test_future_membership_disallows_auto_enroll(self):
+    def test_start_datetime_short_date(self):
+        """verify that the start date for a ccx formats properly by default"""
+        start = datetime(2015, 1, 1, 12, 0, 0, tzinfo=UTC())
+        # This relies on SHORT_DATE remaining the same, is there a better way?
```
Either confirm that tests always run with a fixed SHORT_DATE, or patch SHORT_DATE here so that you know what it is.  The first seems the better option. There are likely other tests with this reliance.

cewing commented:
I appreciate the feedback, @nedbat, @sarina.  Something must be munged on my local instance with `paver run_quality` because I got no error output locally, but can see the reports in the test artifacts from the run above.  I'll polish that stuff up today and get this back to you asap

sarina commented:
Are you using devstack?
On Apr 23, 2015 1:36 PM, "Cris Ewing" <notifications@github.com> wrote:

> I appreciate the feedback, @nedbat <https://github.com/nedbat>, @sarina
> <https://github.com/sarina>. Something must be munged on my local
> instance with paver run_quality because I got no error output locally,
> but can see the reports in the test artifacts from the run above. I'll
> polish that stuff up today and get this back to you asap
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/edx/edx-platform/pull/7771#issuecomment-95666006>.
>

cewing commented:
@sarina I am, yes.

nedbat commented:
On this question: "In particular, we are curious as to the advisability of replicating API provided by a Course Descriptor as part of the CCX model."
Implementing methods and properties to make CustomCourseForEdX act like a true course is a fine approach.  If the new key architecture work obviates it, then fine, and if it doesn't, this is the typical Python approach to these sorts of problems.

nedbat commented on lms/djangoapps/ccx/models.py:
```diff
@@ -16,6 +26,71 @@ class CustomCourseForEdX(models.Model):
+
+    @property
+    def start(self):
+        if self._start is None:
```
BTW, we use a third-party package called lazy to create compute-once properties like this without the boilerplate:
```
@lazy
def start(self):
    from .overrides import get_override_for_ccx
    return get_override_for_ccx(self, self.course, 'start')
```

cewing replied:
Oh, excellent!  I'll rewrite to use that.  Thanks!

cewing commented:
@sarina I believe I've addressed all the comments you and @nedbat made above.  I've also eliminated code quality violations.  All the tests pass here, so this is ready to go, I think.
